modules/default/utils.js

```diff
@@ -17,20 +17,26 @@ async function performWebRequest (url, type = "json", useCorsProxy = false, requ
 		requestUrl = url;
 		request.headers = getHeadersToSend(requestHeaders);
 	}
+
 	const response = await fetch(requestUrl, request);
-	const data = await response.text();
+	if (response.ok) {
+		const data = await response.text();
 
-	if (type === "xml") {
-		return new DOMParser().parseFromString(data, "text/html");
-	} else {
-		if (!data || !data.length > 0) return undefined;
+		if (type === "xml") {
+			return new DOMParser().parseFromString(data, "text/html");
+		} else {
+			if (!data || !data.length > 0) return "null"; //undefined;
```

Comment:
Return `"null"` as a string looks unusual. Can you explain why you did this and why there is a comment `// undefined`?

Reply:
and looking at the provider that calls the performWebRequest (here looking at pirateweather) 
```
	fetchWeatherForecast () {
		this.fetchData(this.getUrl())
			.then((data) => {
				if (!data || !data.daily || !data.daily.data.length) {  // undefined or null (no quotes) 
					// No usable data?
					return;  // <----- I think this bypasses the finally, so the provider is waiting forever, hung
				}

				const forecast = this.generateWeatherObjectsFromForecast(data.daily.data);
				this.setWeatherForecast(forecast);
			})
			.catch(function (request) {   // <--- maybe throw is a better choice
				Log.error("Could not load data ... ", request);
			})
			.finally(() => this.updateAvailable());
```

Reply:
this is the fetchData in the providers layer
```
	async fetchData (url, type = "json", requestHeaders = undefined, expectedResponseHeaders = undefined) {
		const mockData = this.config.mockData;
		if (mockData) {
			const data = mockData.substring(1, mockData.length - 1);
			return JSON.parse(data);
		}
		const useCorsProxy = typeof this.config.useCorsProxy !== "undefined" && this.config.useCorsProxy;
		return performWebRequest(url, type, useCorsProxy, requestHeaders, expectedResponseHeaders, config.basePath);
	}
```